interfaces/builtin/pwm.go

```diff
@@ -0,0 +1,153 @@
+	return "pwm"
+}
+
+func (iface *pwmInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              pwmSummary,
+		BaseDeclarationSlots: pwmBaseDeclarationSlots,
+	}
+}
+
+// BeforePrepareSlot checks the slot definition is valid
+func (iface *pwmInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	// must have a PWM chanel number
+	number, ok := slot.Attrs["number"]
+	if !ok {
+		return fmt.Errorf("pwm slot must have a number attribute")
+	}
+
+	// valid values of number
+	if _, ok := number.(int64); !ok {
+		return fmt.Errorf("pwm slot number attribute must be an int")
+	}
+
+	// must have a PWM chip number
+	chipNum, ok := slot.Attrs["chip-number"]
+	if !ok {
+		return fmt.Errorf("pwm slot must have a chip-number attribute")
+	}
+
+	// valid values of chip number
+	if _, ok := chipNum.(int64); !ok {
+		return fmt.Errorf("pwm slot chip-number attribute must be an int")
+	}
+
+	// slot is good
+	return nil
+}
+
+func (iface *pwmInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var chipNum int64
+	if err := slot.Attr("chip-number", &chipNum); err != nil {
+		return err
+	}
+
+	var number int64
+	if err := slot.Attr("number", &number); err != nil {
+		return err
+	}
+	path := fmt.Sprintf(pwmSysfsPwmChipBase, chipNum)
+	// Entries in /sys/class/pwm for PWM chips are just symlinks
+	// to their correct device part in the sysfs tree. Given AppArmor
+	// requires symlinks to be dereferenced, evaluate the PWM
+	// path and add the correct absolute path to the AppArmor snippet.
+	dereferencedPath, err := evalSymlinks(path)
+	if err != nil && os.IsNotExist(err) {
+		// If the specific pwm is not available there is no point
+		// exporting it, we should also not fail because this
+		// will block snapd updates (LP: 1866424)
+		logger.Noticef("cannot find not existing pwm chipbase %s", path)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	spec.AddSnippet(fmt.Sprintf("%s/pwm%d/* rwk,", dereferencedPath, number))
+	return nil
+}
+
+func (iface *pwmInterface) SystemdConnectedSlot(spec *systemd.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var chipNum int64
+	if err := slot.Attr("chip-number", &chipNum); err != nil {
+		return err
+	}
+
+	var pwmNum int64
+	if err := slot.Attr("number", &pwmNum); err != nil {
+		return err
+	}
+
+	serviceName := interfaces.InterfaceServiceName(slot.Snap().InstanceName(), fmt.Sprintf("pwmchip%d-pwm%d", chipNum, pwmNum))
+	service := &systemd.Service{
+		Type:            "oneshot",
+		RemainAfterExit: true,
+		ExecStart:       fmt.Sprintf("/bin/sh -c 'test -e /sys/class/pwm/pwmchip%d/pwm%d || echo %d > /sys/class/pwm/pwmchip%d/export'", chipNum, pwmNum, pwmNum, chipNum),
+		ExecStop:        fmt.Sprintf("/bin/sh -c 'test ! -e /sys/class/pwm/pwmchip%d/pwm%d || echo %d > /sys/class/pwm/pwmchip%d/unexport'", chipNum, pwmNum, pwmNum, chipNum),
```

Comment:
more readable versions of these would be:

```suggestion
		ExecStart:       fmt.Sprintf("/bin/sh -c 'test -e /sys/class/pwm/pwmchip%[1]d/pwm%[2]d || echo %[2]d > /sys/class/pwm/pwmchip%[1]d/export'", chipNum, pwmNum),
		ExecStop:        fmt.Sprintf("/bin/sh -c 'test ! -e /sys/class/pwm/pwmchip%[1]d/pwm%[2]d || echo %[2]d > /sys/class/pwm/pwmchip%[1]d/unexport'", chipNum, pwmNum),
```

though the same comment applies to the existing gpio interface too

Reply:
Sure thing!

Reply:
Updated as well.